### ISSUE TITLE
Extremely frequent "graphics card ran out of memory" crashes

### ISSUE THREAD

**Issue by Ryminer2** [anegostudios/VintageStory-Issues#3428]
### Game Version

1.18.15

### Platform

Windows

### Modded

Vanilla

### SP/MP

Singleplayer

### Description

I've played about five hours of this game. The first four were completely without issue. Now, suddenly, when I load into my world, I usually have a few seconds at most before this crash happens. On rare occasions the game remains stable enough for me to play for a while, but there's no consistent pattern to it. I saw that another issue report mentioned this being caused by minimizing the game or tabbing out of it, so I tried to avoid that -- it doesn't make a difference.

The first time this crash occurred, I lowered my graphics settings from High to Medium, thinking that maybe that would lessen the issue, but it's only gotten worse since then. My GPU is not hurting for memory in any way, it's a 3060; I really can't fathom why this crash is happening.

### How to reproduce

1. Have my machine, I guess
2. Load into world
3. ???
4. Crash be upon ye

### Logs

```Log
Running on 64 bit Windows with 16328 MB RAM
Game Version: v1.18.15 (Stable)
1/12/2024 2:54:23 PM: Critical error occurred
Loaded Mods: game@1.18.15, creative@1.18.15, survival@1.18.15
System.OutOfMemoryException: Your graphics card ran out of memory! Please close other programs and reduce your view distance to prevent the game from crashing.
   at Vintagestory.Client.NoObf.ClientPlatformWindows.CheckGlErrorAlways(String errmsg) in VintagestoryLib\Client\ClientPlatform\GLImpl.cs:line 85
   at Vintagestory.Client.NoObf.ClientMain.RenderAfterFinalComposition(Single dt) in VintagestoryLib\Client\ClientMain.cs:line 931
   at Vintagestory.Client.GuiScreenRunningGame.RenderAfterFinalComposition(Single dt) in VintagestoryLib\Client\MainMenu\Screens\GuiScreenRunningGame.cs:line 219
   at Vintagestory.Client.ScreenManager.Render(Single dt) in VintagestoryLib\Client\ScreenManager.cs:line 668
   at Vintagestory.Client.ScreenManager.OnNewFrame(Single dt) in VintagestoryLib\Client\ScreenManager.cs:line 643
   at Vintagestory.Client.NoObf.ClientPlatformWindows.window_RenderFrame(FrameEventArgs e) in VintagestoryLib\Client\ClientPlatform\GameWindow.cs:line 77
   at OpenTK.Windowing.Desktop.GameWindow.Run()
   at Vintagestory.Client.ClientProgram.Start(ClientProgramArgs args, String[] rawArgs) in VintagestoryLib\Client\ClientProgram.cs:line 313
   at Vintagestory.Client.ClientProgram.<>c__DisplayClass9_0.<.ctor>b__1() in VintagestoryLib\Client\ClientProgram.cs:line 129
   at Vintagestory.ClientNative.CrashReporter.Start(ThreadStart start) in VintagestoryLib\Client\ClientPlatform\ClientNative\CrashReporter.cs:line 93

Event Log entries containing Vintagestory.exe, the latest 3
==================================
{ TimeGenerated = 1/12/2024 2:53:26 PM, Site = , Source = Windows Error Reporting, Message = Fault bucket 2007150627803250050, type 4
Event Name: APPCRASH
Response: Not available
Cab Id: 0

Problem signature:
P1: Vintagestory.exe
P2: 1.18.15.0
P3: 64e841c5
P4: openal32.dll
P5: 1.23.0.0
P6: 63dd31ad
P7: 40000015
P8: 00000000000df046
P9: 
P10: 

Attached files:
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WERACFB.tmp.WERInternalMetadata.xml

These files may be available here:
\\?\C:\ProgramData\Microsoft\Windows\WER\ReportArchive\AppCrash_Vintagestory.exe_3278c09e339ad58fdb4a48929ddf95e5f2060ca_ee8069fc_cebe146e-6b12-4fa3-a4f8-209b2e030558

Analysis symbol: 
Rechecking for solution: 0
Report Id: 128487a5-dfe9-4fd3-ac79-7d20d0da6f11
Report Status: 268435456
Hashed bucket: 38dd2bbe130a69d3fbdad4dc94ea3d82
Cab Guid: 0 }
--------------
{ TimeGenerated = 1/12/2024 2:53:26 PM, Site = , Source = Windows Error Reporting, Message = Fault bucket , type 0
Event Name: APPCRASH
Response: Not available
Cab Id: 0

Problem signature:
P1: Vintagestory.exe
P2: 1.18.15.0
P3: 64e841c5
P4: openal32.dll
P5: 1.23.0.0
P6: 63dd31ad
P7: 40000015
P8: 00000000000df046
P9: 
P10: 

Attached files:

These files may be available here:
\\?\C:\ProgramData\Microsoft\Windows\WER\ReportQueue\AppCrash_Vintagestory.exe_3278c09e339ad58fdb4a48929ddf95e5f2060ca_ee8069fc_cebe146e-6b12-4fa3-a4f8-209b2e030558

Analysis symbol: 
Rechecking for solution: 0
Report Id: 128487a5-dfe9-4fd3-ac79-7d20d0da6f11
Report Status: 4
Hashed bucket: 
Cab Guid: 0 }
--------------
{ TimeGenerated = 1/12/2024 2:53:24 PM, Site = , Source = Application Error, Message = Faulting application name: Vintagestory.exe, version: 1.18.15.0, time stamp: 0x64e841c5
Faulting module name: openal32.dll, version: 1.23.0.0, time stamp: 0x63dd31ad
Exception code: 0x40000015
Fault offset: 0x00000000000df046
Faulting process id: 0x193a4
Faulting application start time: 0x01da4590e72a96e1
Faulting application path: C:\Users\Riley\AppData\Roaming\Vintagestory\Vintagestory.exe
Faulting module path: C:\Users\Riley\AppData\Roaming\Vintagestory\Lib\openal32.dll
Report Id: 128487a5-dfe9-4fd3-ac79-7d20d0da6f11
Faulting package full name: 
Faulting package-relative application ID:  }
```

**Comment by Ryminer2** [comment]
![gpu explosion](https://github.com/anegostudios/VintageStory-Issues/assets/6020841/eef57cdf-9aaf-4254-a3a1-5b292302f8d3)
This is what my graphics memory does when I try to load into my world. Not really a memory "leak" so much as a memory "explosion." Most of the time it crashes my whole video card.

**Comment by tyronx** [comment]
hi, is this crash specific to one savegame or it happens on all savegames / multiplayer servers?

**Comment by radfast** [comment]
Please try a couple of things:

- try game version 1.19.1 released today, it fixes hundreds of issues in 1.18.15, though not specifically this issue it might have been fixed indirectly/incidentally
- before starting the game, in the Main Menu / Settings menu, try substantially lowering your view distance to something like 128.

If view distance 128 works OK, then try increasing it to something which is comfortable for you but does not cause the issue.

**Comment by Artalus** [comment]
I can confirm that this problem (or something extremely similar to it) is present in `1.19.1` on a fresh world with all mods disabled. I started experiencing frequent black screen flickering and out-of-memory crashes on our modded server after updating to 1.19.1, and my friend had fps drops that were fixed only by restarting the whole game, so I decided to investigate if there are were memory leaks at play in Vanilla as well.

I drafted a small powershell script to dump `VintageStory.exe`'s private working set size and started loading the same (newly created) world, exiting to main menu, and reloading again. On the 9th reload VS crashed with almost the same message that OP had.

<details><summary>Log</summary>
<p>

```
Running on 64 bit Windows 10.0.19044.0 with 65467 MB RAM
Game Version: v1.19.1 (Stable)
2024-01-21 15:59:32: Critical error occurred
Loaded Mods: game@1.19.1, creative@1.19.1, survival@1.19.1
System.OutOfMemoryException: Either the graphics card or the OS ran out of memory! Please close other programs and reduce your view distance to prevent the game from crashing.
   at Vintagestory.Client.NoObf.ClientPlatformWindows.CheckGlErrorAlways(String errmsg) in VintagestoryLib\Client\ClientPlatform\GLImpl.cs:line 75
   at Vintagestory.Client.ScreenManager.Render(Single dt) in VintagestoryLib\Client\ScreenManager.cs:line 675
   at Vintagestory.Client.ScreenManager.OnNewFrame(Single dt) in VintagestoryLib\Client\ScreenManager.cs:line 650
   at Vintagestory.Client.NoObf.ClientPlatformWindows.window_RenderFrame(FrameEventArgs e) in VintagestoryLib\Client\ClientPlatform\GameWindow.cs:line 78
   at OpenTK.Windowing.Desktop.GameWindow.Run()
   at Vintagestory.Client.ClientProgram.Start(ClientProgramArgs args, String[] rawArgs) in VintagestoryLib\Client\ClientProgram.cs:line 318
   at Vintagestory.Client.ClientProgram.<>c__DisplayClass9_0.<.ctor>b__1() in VintagestoryLib\Client\ClientProgram.cs:line 128
   at Vintagestory.ClientNative.CrashReporter.Start(ThreadStart start) in VintagestoryLib\Client\ClientPlatform\ClientNative\CrashReporter.cs:line 93

Event Log entries for Vintagestory.exe, the latest 1
==================================
{ TimeGenerated = 2024-01-20 19:11:18, Site = , Source = Application Error, Message = Faulting application name: Vintagestory.exe, version: 1.19.1.0, time stamp: 0x65310000
Faulting module name: openal32.dll, version: 1.23.0.0, time stamp: 0x63dd31ad
Exception code: 0x40000015
Fault offset: 0x00000000000df046
Faulting process id: 0x4c48
Faulting application start time: 0x01da4ba2505f9dce
Faulting application path: E:\games\Vintagestory\Vintagestory.exe
Faulting module path: E:\games\Vintagestory\Lib\openal32.dll
Report Id: ee8efcdc-1162-4d0c-86f4-b3444df5499b
Faulting package full name:
Faulting package-relative application ID:  }
```

</p>
</details> 

Here is the memory consumption graph that I measured at that moment:

![1](https://github.com/anegostudios/VintageStory-Issues/assets/10381253/48e048a9-1a36-4a1f-a3dd-bb937ba1161e)

`0.1` is just the menu loaded, `0.2` is world generation completed with a character sellection screen, `1` is a world fully loaded for the first time. Note that each subsequnt reload eats slightly more RAM up until `5`.
1 through 5 I was waiting for the memory to cleanup in the menu before reloading; on `6` through `9` I reloaded immediately once I got to the menu. Notice how VS takes more memory and does not release it until you go back to the menu.
Reload `9` resulted in both my displays going black, and a crash reporter window with the log from above.

(note that I was measuring the WorkingSetPrivate size, which reports lower values than PrivateBytes or WorkingSetSize often used by other memory diagnostic tools)

Subsequent experiments give me mixed readings. The game does behave weirdly on subsequent reloads, with RAM spiking up to 10+ gigs and VRAM to 3+; it doesn't *always* cause a crash, but it always causes at least a ~1 second freeze. Sometimes it is not the game that crashes, but one of the others running applications - Visual Studio Code, Discord, StreamDeck etc. It does feel as if the system is running out of not RAM or VRAM, but of some other resource.

The situation worsens significantly if I use the `Rivers` mod and reload a world (newly created as well) with a river running through it, but I am unsure if this is related.

---

My specs&settings are:
- Windows 10 x64
- 64 GB RAM (at least 50% is always free at the moment of a glitch happening)
- RTX 3060 (12GB VRAM)
- View distance 512
- Optimize RAM: somewhat
- VS installed on HDD, system & saves located on M.2 NVME; lots of free space on both.

**Comment by james-berkheimer** [comment]
I'm encountering a similar issue as @Artalus.  Same graphics card but with only 32GB of RAM.  The GPU memory slowly climbs until the game crashes and any other program that's relying on the GPU crashes as well (mainly my Chrome browser).  I'm also running 1.19.3.

**Comment by Craluminum2413** [comment]
#1712

**Comment by Craluminum2413** [comment]
Closing due to inactivity
